src/crush/mapper.c

```diff
@@ -776,6 +885,69 @@ static void crush_choose_indep(const struct crush_map *map,
 #endif
 }
 
+
+/* This takes a chunk of memory and sets it up to be a shiny new
+   working area for a CRUSH placement computation. In principal we
+   could do less work on re-use of a previous area, but since that
+   would still involve looping over all buckets and conditional
+   behavior on each, it's not worth it. */
+
+static void reset_work(const struct crush_map* m, struct crush_work* w) {
+	/* We work by moving through the available space and setting
+	   values and pointers as we go.
+
+	   It's a bit like Forth's use of the 'allot' word since we
+	   set the pointer first and then reserve the space for it to
+	   point to by incrementing the cursor. */
+	char* point = (char*)w;
+	point += sizeof(struct crush_work*);
+	w->work = (struct crush_work_bucket**) point;
+	point += m->max_buckets * sizeof(struct crush_work_bucket*);
+	for (size_t b = 0; b < m->max_buckets; ++b) {
+		if (m->buckets[b] == 0)
+			continue;
+
+		w->work[b] = (struct crush_work_bucket*) point;
+		switch (m->buckets[b]->alg) {
+		case CRUSH_BUCKET_STRAW2:
+			;
+			struct crush_work_straw2* s2w
+				= (struct crush_work_straw2 *) point;
+			point += sizeof(struct crush_work_straw2);
+			struct crush_bucket_straw2* s2b
+				= (struct crush_bucket_straw2 *) m->buckets[b];
+			s2w->weight_sum = 0;
+			s2w->picked_sum = 0;
+			for (size_t i = 0; i < s2b->h.size; ++i)
+				s2w->weight_sum += s2b->item_weights[i];
+
+			break;
+		case CRUSH_BUCKET_LIST:
+			point += sizeof(struct crush_work_list);
```

Comment:
In the straw2 case you assign a straw2 pointer and then advance over the straw2 struct, whereas here you skip the list struct and then assign a list pointer.  I haven't followed the memory layout, so it might be OK, but it stands out.